exec/Chainweb/Gaps.hs

```diff
@@ -52,79 +46,54 @@ gaps env args = do
+        let doChain (cid, gs) = do
+              let ranges = concatMap (createRanges cid) gs
+              mapM_ (f logg count cid) ranges
         let gapFiller = do
-              race_ (progress logg count totalNumBlocks) $
-                traverseMapConcurrently_ Par' (\cid -> traverseConcurrently_ strat (f logg blockQueue count cid) . concatMap (createRanges cid)) gapsByChain
+              race_ (progress logg count totalNumBlocks)
+                    (traverseConcurrently_ Par' doChain (M.toList gapsByChain))
               final <- readIORef count
               logg Info $ fromString $ printf "Filled in %d missing blocks." final
         if disableIndexesPred
           then withDroppedIndexes pool logg gapFiller
           else gapFiller
-        flushQueue blockQueue count
-        final <- readIORef count
-        logg Info $ fromString $ printf "Filled in %d missing blocks." final
   where
     pool = _env_dbConnPool env
     delay =  _gapArgs_delayMicros args
     disableIndexesPred =  _gapArgs_disableIndexes args
     gi = mkGenesisInfo $ _env_nodeInfo env
     logg = _env_logger env
-    traverseMapConcurrently_ comp g m =
-      withScheduler_ comp $ \s -> scheduleWork s $ void $ M.traverseWithKey (\k -> scheduleWork s . void . g k) m
     createRanges cid (low, high)
       | low == high = []
       | fromIntegral (genesisHeight (ChainId (fromIntegral cid)) gi) == low = rangeToDescGroupsOf blockHeaderRequestSize (Low $ fromIntegral low) (High $ fromIntegral (high - 1))
       | otherwise = rangeToDescGroupsOf blockHeaderRequestSize (Low $ fromIntegral (low + 1)) (High $ fromIntegral (high - 1))
-    flushQueue :: TBQueue (Vector BlockHeader) -> IORef Int -> IO ()
-    flushQueue blockQueue count =
-      atomically (tryReadTBQueue blockQueue) >>= \case
-        Nothing -> logg Debug "blockQueue fully flushed."
-        Just q -> do
-          writeBlocks env pool disableIndexesPred count (V.toList q)
-          flushQueue blockQueue count
 
-    f :: LogFunctionIO Text -> TBQueue (Vector BlockHeader) -> IORef Int -> Int64 -> (Low, High) -> IO ()
-    f logger blockQueue count cid (l, h) = do
+    f :: LogFunctionIO Text -> IORef Int -> Int64 -> (Low, High) -> IO ()
+    f logger count cid (l, h) = do
       let range = (ChainId (fromIntegral cid), l, h)
+      --logger Debug $ fromString $ printf "Processing range %s" (show range)
       headersBetween env range >>= \case
         Left e -> logger Error $ fromString $ printf "ApiError for range %s: %s" (show range) (show e)
         Right [] -> logger Error $ fromString $ printf "headersBetween: %s" $ show range
-        Right hs -> do
-          let vs = V.fromList hs
-          mq <- atomically (tryReadTBQueue blockQueue)
-          maybe mempty (writeBlocks env pool disableIndexesPred count . V.toList) mq
-          atomically $ writeTBQueue blockQueue vs
+        Right hs -> writeBlocks env pool disableIndexesPred count hs
```

Comment:
Each thread processes its gaps sequentially, eliminating the need for the queue.